File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int[] vis = new int[26];
        int[] freq = new int[26];

        for(char ch: s.toCharArray()){
            freq[ch-'a']++;
        }

        Stack<Character> st = new Stack<>();

        for(char ch : s.toCharArray()){
            freq[ch - 'a']--;
            if(vis[ch - 'a'] == 1){
                continue;
            }
            while(!st.empty() && st.peek() > ch && freq[st.peek() - 'a'] > 0){
                vis[st.pop() - 'a'] = 0;
            }
            st.push(ch);
            vis[ch - 'a'] = 1;
        }

        StringBuilder ans = new StringBuilder();
        for(char c : st){
            ans.append(c);
        }

        return ans.toString();
    }
}

File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int> freq(26, 0);
        vector<int> seen(26, 0);
        for(int ch:s){
            freq[ch - 'a']++;
        }
        stack<char> st;
        for(char ch : s) {
            freq[ch - 'a']--;
            if(seen[ch - 'a']) continue;
            while(!st.empty() && st.top() > ch && freq[st.top() - 'a'] > 0) {
                seen[st.top() - 'a'] = false;
                st.pop();
            }
            st.push(ch);
            seen[ch-'a'] = true;
        }
        string ans ;
        while(!st.empty()) {
            ans += st.top();
            st.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;        
    }
};